### PR TITLE
Clean consumer stats samples differently depending on how many there are

### PR DESCRIPTION
**Pull request by SimonUnge**
## Proposed Changes

This is a revision of https://github.com/rabbitmq/rabbitmq-server/pull/9356.

Check number of consumers for a terminated channel, If over 1000, bulk delete metrics, if under 1001, do it the old way.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [x] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [x] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.

### CODE REVIEW

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl [hunk 1]
```diff
@@ -82,13 +85,13 @@ remove_connection(Id, BIntervals) ->
     delete_samples(connection_stats_coarse_conn_stats, Id, BIntervals),
     ok.
 
-remove_channel(Id, BIntervals) ->
+remove_channel(Id, CMCount, BIntervals) ->
```
Comment (michaelklishin) [inline deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:88]: Any reason not make it a few characters longer and say `ConsumerCount`?

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl [hunk 2]
```diff
@@ -82,13 +85,13 @@ remove_connection(Id, BIntervals) ->
...
     ets:delete(channel_created_stats, Id),
     ets:delete(channel_stats, Id),
     delete_samples(channel_process_stats, Id, BIntervals),
     delete_samples(channel_stats_fine_stats, Id, BIntervals),
     delete_samples(channel_stats_deliver_stats, Id, BIntervals),
-    index_delete(consumer_stats, channel, Id),
+    index_delete(consumer_stats, {channel, CMCount}, Id),
```
Comment (michaelklishin) [inline deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:94]: Same as above

File: deps/rabbit/src/rabbit_channel.erl [hunk 3]
```diff
@@ -824,8 +824,10 @@ terminate(_Reason,
                   rabbit_global_counters:consumer_deleted(amqp091)
           end, CM),
     rabbit_core_metrics:channel_closed(self()),
+    CMSize = maps:size(CM),
     rabbit_event:notify(channel_closed, [{pid, self()},
-                                         {user_who_performed_action, Username}]),
+                                         {user_who_performed_action, Username},
+                                         {consumer_count, CMSize}]),
```
Comment (michaelklishin) [inline deps/rabbit/src/rabbit_channel.erl:830]: Same as below

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl [hunk 4]
```diff
@@ -137,18 +140,23 @@ delete_samples(Table, Id, Intervals) ->
...
-    %% Delete consumer_stats_queue_index
-    ets:match_delete(consumer_stats_queue_index,
-                     {'_', MatchPattern}),
-    %% Delete consumer_stats
-    ets:match_delete(consumer_stats,
-                     {MatchPattern,'_'}),
-    %% Delete consumer_stats_channel_index
-    ets:delete(IndexTable, Id),
-    ok;
+index_delete(consumer_stats = Table, {channel = Type, Count}, Id) ->
```
Comment (michaelklishin) [inline deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:143]: I'd also be specific and make it say `ConsumerCount` here, too

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl [hunk 5]
```diff
@@ -137,18 +140,23 @@ delete_samples(Table, Id, Intervals) ->
...
+    case Count > ?LARGE_CONSUMER_COUNT of
```
Comment (michaelklishin) [inline deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:144]: We can add a comment explaining the condition here: `ets:match_delete/2` is acceptable or optimal for a large number of entries **on this particular channel** but when there are many channels with a few (very often just one) consumers each, its sequential scans become prohibitively expensive.

**Review by michaelklishin — CHANGES_REQUESTED** [review]

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl [hunk 6]
```diff
@@ -137,18 +140,23 @@ delete_samples(Table, Id, Intervals) ->
...
+            %% Delete consumer_stats
+            ets:match_delete(consumer_stats,
+                             {MatchPattern,'_'}),
+            %% Delete consumer_stats_channel_index
+            ets:delete(IndexTable, Id),
+            ok;
+        false ->
+            index_delete(Table, Type, Id)
+    end;
 index_delete(Table, Type, Id) ->
```
Comment (michaelklishin) [inline deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:160]: Just for completeness' sake, if a node gets an event that does not have the 2nd argument as a tuple, this function head should handle that. E.g. in a mixed version cluster situation during an ongoing upgrade.

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl (hunk 1 above)
Comment (SimonUnge) [reply deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:88]: No valid reason! I guess my C days still are influencing my var names.

File: deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl (hunk 6 above)
Comment (SimonUnge) [reply deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:160]: Hm, but if it does not have a second argument, i.e Type = consumer, line 160 should handle it, the way it did before I introduced the match_delete logic?

Comment (SimonUnge) [reply deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:164]: @michaelklishin fixed the var names and comment. I think I am missing something here though!

Comment (michaelklishin) [reply deps/rabbitmq_management_agent/src/rabbit_mgmt_metrics_gc.erl:164]: Yes, that's what I was referring to. Like 160 would handle the case without the consumer count :)

**Comment by michaelklishin** [comment]
The forced push was a rebase.

**Review by kjnilsson — APPROVED** [review]
LGTM! Simpler than the idea I had but should be effective. Cheers.

**Comment by SimonUnge** [comment]
> LGTM! Simpler than the idea I had but should be effective. Cheers.

@kjnilsson Curious, what was your idea?

**Comment by michaelklishin** [comment]
Merged with a cosmetic change from me in https://github.com/rabbitmq/rabbitmq-server/pull/10476

**Pull request closed without merging**